src/Microsoft.AspNet.SignalR.Core/Infrastructure/DispatchingTaskCompletionSource.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.SignalR.Infrastructure
+{
+    internal class DispatchingTaskCompletionSource<TResult>
+    {
+        private readonly TaskCompletionSource<TResult> _tcs = new TaskCompletionSource<TResult>();
+
+        public Task<TResult> Task
+        {
+            get { return _tcs.Task; }
+        }
+
+        public void SetCanceled()
```

Comment:
Not at the moment. I just tried to make it as similar to TaskCompletionSource as possible. We might need it when replacing TaskCompletionSources used on the server side. I will leave it for now and will replace TaskCompletionSources on the server side with this one and then remove methods that are not used/needed.